### DISCUSSION TITLE
How do we setup Jest for unit testing?

### DISCUSSION THREAD

**Discussion by lanegoolsby** [facebook/docusaurus#7234] [General]
I've been trying to setup Jest. I am unable to get it working though.

The problem I cannot get past is getting Jest to recognize the Docusaurus modules. I've attached a minimum reproduction of the issue. I know its a configuration issue but I've come at this problem dozens of different directions all to no avail. 

[unit-tests.zip](https://github.com/facebook/docusaurus/files/8548628/unit-tests.zip)

![image](https://user-images.githubusercontent.com/10758510/164946315-201aae72-fa94-4288-a0ef-8087900cc295.png)

**Comment by slorber** [comment]
On your Jest config, maybe adding a way for Jest to resolve the Webpack alias modules can help fix this. We do use this in our own repo config.

For example, you can try:
  
```js
{
  moduleNameMapper: {
    '@docusaurus/(BrowserOnly|ComponentCreator|constants|ExecutionEnvironment|Head|Interpolate|isInternalUrl|Link|Noop|renderRoutes|router|Translate|use.*)': '@docusaurus/core/src/client/exports/$1', 
    '@theme/(.*)': '@docusaurus/theme-classic/src/theme/$1',
    '@site/(.*)': 'website/$1',
  }
}
```


Note: the `@docusaurus/xyz` config will likely become simpler with TS 4.7 + Jest 28 supporting package `exports`

**Reply by lanegoolsby** [reply]
Same error. Here is my jest.config.js. Note that I changed the path for @docusaurus/* to be `@docusaurus/core/lib/client/exports/$1` (swapped /src for /lib since /src does not exist in my node_modules folder). The path for the @theme mapper did have the src folder.

```
module.exports = {
    preset: 'ts-jest',
    transform: {
        '^.+\\.{ts,tsx}?$': 'ts-jest'
    },
    moduleNameMapper: {
        '@docusaurus/(BrowserOnly|ComponentCreator|constants|ExecutionEnvironment|Head|Interpolate|isInternalUrl|Link|Noop|renderRoutes|router|Translate|use.*)': '@docusaurus/core/lib/client/exports/$1', 
        '@theme/(.*)': '@docusaurus/theme-classic/src/theme/$1',
        '@site/(.*)': 'website/$1',
      }
};
```

![image](https://user-images.githubusercontent.com/10758510/165960306-d82bef0f-09fc-4d2c-b8e5-ee1818007c63.png)

![image](https://user-images.githubusercontent.com/10758510/165960452-8e343f22-c6d3-4548-b6be-9c71c98dfd8d.png)

**Reply by slorber** [reply]
It is surprising to have the exact same error, are you sure?

You'll also need this to load CSS files and other assets:

```
    '^.+\\.(css|jpe?g|png|svg|webp)$': '<rootDir>/jest/emptyModule.ts',
```

(requires creating emptyModule.ts file)

this is what we use on our own repo, maybe there's some mapping missing 🤷‍♂️ 

source is https://github.com/facebook/docusaurus/blob/main/jest.config.mjs#L50

**Reply by tuckergordon** [reply]
I've also been running into issues trying to get Jest working with Docusaurus+TypeScript. I forked the Sandbox Stackblitz and used the above to add a simple Jest test to it: [https://stackblitz.com/edit/github-yjetx2?file=jest.config.js](https://stackblitz.com/edit/github-yjetx2?file=jest.config.js).

A few key changes I had to make:
- Instead of using `ts-jest`, import `@swc/jest` and use the `transform` listed above. That seemed to resolve most of the errors I was getting
- Adding the `moduleNameMapper` listed above as well as the `emptyModule.ts` file

This _almost_ works except it's taking issue with the `Svg` component in `HomepageFeatures/index.tsx`. If you comment out line 49, the test passes:

```tsx
{/* <Svg className={styles.featureSvg} role="img" /> */}
```

I'm going to keep trying to debug this, though I have some lingering thoughts/concerns:

- This Svg issue could be indicative of a more fundamental issue that doesn't pop up elsewhere in this simple example
- It would be nice if running jest could rely on the same compilation process that `npm run build` uses, rather than needing to introduce `swc` (or `ts-jest`)
- Once this is all figured out, the docs could definitely benefit from an example and description of how to setup Docusaurus + jest

**Reply by lanegoolsby** [reply]
Could the SVG issue you're having be related to [this](https://github.com/facebook/docusaurus/issues/9715)? That's giving us a lot of problems in our build pipelines where sometimes the build passes, sometimes it doesn't (but that may be our own internal ish). If that's not unique to our environment that could be causing merry chaos with jest as well.

**Comment by karl-cardenas-coding** [comment]
I, too, have spent the majority of the day trying to create a test case for a custom component that uses Docusarus dependencies.  I tried the suggestions in this thread, but to no avail. Something interesting I noticed is that no one is testing their custom components. I took a glance at the showcase page, and none of the repositories I visited had test cases for custom components. 

Are we, as a community, not testing our custom components? Is it because of the challenges we are facing? 
¯\_(ツ)_/¯,

**Reply by slorber** [reply]
@karl-cardenas-coding we test a bit our theme components in Docusaurus packages, but could improve there.

To be honest, unless you build very interactive components (not the majority in docs sites), is it really useful to test them with Jest? Asserting a div as a given CSS rule or text doesn't really prevent most common regressions. I much prefer testing with visual regression tests, as explained in https://docusaurus.io/blog/upgrading-frontend-dependencies-with-confidence-using-visual-regression-testing

Now, if you have components you want to test in Jest and don't know how, please provide a repro and I'll take a look. It really depends on what you are testing and the dependencies used. In any case you can mock/alias dependencies in Jest, or use dependency injection if really needed. I can only help if I see/run your code, so producing a https://docusaurus.new/stackblitz repro sandbox is always helpful.

**Comment by karl-cardenas-coding** [comment]
@slorber First of all thank you so much for the detailed blog. We implemented a very similar solution, except we went with Playwright to do the comparison versus using Argos.  Instead of storing the images in git, we use GitHub Artifacts. The report is posted on a GitHub page - we create a unique URL per test run. 

Thank you for sharing this with us. This definitely goes a long way in helping us feel more confident about rolling out changes 💪🏻

**Comment by zhangjianrencai** [comment]
@lanegoolsby @slorber any update about this discussion. I met the same issue that I use jest to mock `@docusaurus/**`. and I also add 

`module.exports = {
    testEnvironment: "jsdom",
    transform: {
      '^.+\\.[jt]sx?$': [
        '@swc/jest',
        {
          jsc: {
            parser: {
              syntax: 'typescript',
              tsx: true,
            },
            target: 'es2020',
          },
        },
      ],
    },
    moduleNameMapper: {
      // Jest can't resolve CSS or asset imports
      "^.+\\.(css|jpe?g|png|svg|webp)$": "<rootDir>/jest/emptyModule.ts",
      '@docusaurus/(BrowserOnly|ComponentCreator|constants|ExecutionEnvironment|Head|Interpolate|isInternalUrl|Link|Noop|renderRoutes|router|Translate|use.*)': '@docusaurus/core/lib/client/exports/$1', 
      '@theme/(.*)': '@docusaurus/theme-classic/src/theme/$1',
      '@site/(.*)': 'website/$1',
    },
  };
  `
  However,it told us 
  
![image](https://github.com/user-attachments/assets/3e0c7402-7653-4f0c-9d70-3e99f6c9e198)
